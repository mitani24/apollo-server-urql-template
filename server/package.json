{
  "name": "server",
  "version": "1.0.0",
  "license": "MIT",
  "scripts": {
    "dev": "ts-node-dev src/index.ts",
    "start": "ts-node src/index.ts",
    "mock:api": "prism mock ../openapi/openapi.yaml"
  },
  "dependencies": {
    "apollo-datasource-rest": "^3.1.1",
    "apollo-server": "^3.1.2",
    "graphql": "^15.5.1"
  },
  "devDependencies": {
    "@stoplight/prism-cli": "^4.3.2",
    "@types/jest": "^27.0.1",
    "jest": "^27.0.6",
    "ts-node": "^10.2.0",
    "ts-node-dev": "^1.1.8",
    "typescript": "^4.3.5"
  }
}
