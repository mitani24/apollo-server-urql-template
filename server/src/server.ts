import fs from "fs";
import path from "path";
import { ApolloServer, gql } from "apollo-server";
import TrackAPI from "./datasources/track-api";
import resolvers from "./resolvers";
import { ApolloServerExpressConfig } from "apollo-server-express";

const typeDefs = fs
  .readFileSync(path.join(__dirname, "../../graphql/schema.graphql"))
  .toString();

export const createServer = (
  context?: ApolloServerExpressConfig["context"]
) => {
  return new ApolloServer({
    typeDefs: gql`
      ${typeDefs}
    `,
    dataSources() {
      return { trackAPI: new TrackAPI() };
    },
    resolvers,
    context,
  });
};
