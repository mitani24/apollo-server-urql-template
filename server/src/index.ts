import { createServer } from "./server";

const server = createServer();

server.listen().then(() => {
  console.log(`
    ğŸš€  Server is running!
    ğŸ”‰  Listening on port 4000
    ğŸ“­  Query at https://studio.apollographql.com/dev
  `);
});
