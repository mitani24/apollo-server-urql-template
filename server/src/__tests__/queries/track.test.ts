import { gql } from "apollo-server-core";
import { createServer } from "../../server";

const TRACK = gql`
  query track($trackId: ID!) {
    track(id: $trackId) {
      id
      title
      author {
        id
        name
        photo
      }
      thumbnail
      length
      modulesCount
      description
      numberOfViews
      modules
    }
  }
`;

describe("track", () => {
  it("fetches the track specified by trackId", async () => {
    const server = createServer();
    const res = await server.executeOperation({
      query: TRACK,
      variables: { trackId: "c_0" },
    });
    expect(res).toMatchSnapshot();
  });

  it("responds with a 404 error", async () => {
    const server = createServer({ mockResponseCode: 404 });
    const res = await server.executeOperation({
      query: TRACK,
      variables: { trackId: "c_0" },
    });
    expect(res).toMatchSnapshot();
  });
});
