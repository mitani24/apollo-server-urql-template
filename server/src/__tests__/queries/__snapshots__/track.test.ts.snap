// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`track fetches the track specified by trackId 1`] = `
Object {
  "data": Object {
    "track": Object {
      "author": Object {
        "id": "cat-i",
        "name": "Yoshi Ikuzo",
        "photo": "https://images.unsplash.com/photo-1442291928580-fb5d0856a8f1?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=1080&fit=max&ixid=eyJhcHBfaWQiOjExNzA0OH0",
      },
      "description": "Cat is God.",
      "id": "c_0",
      "length": 2377,
      "modules": Array [
        "l_0",
        "l_1",
        "l_2",
        "l_3",
        "l_4",
        "l_5",
        "l_6",
        "l_7",
        "l_8",
        "l_9",
      ],
      "modulesCount": 10,
      "numberOfViews": 10,
      "thumbnail": "https://res.cloudinary.com/dety84pbu/image/upload/v1598465568/nebula_cat_djkt9r.jpg",
      "title": "Cat-stronomy, an introduction",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`track responds with a 404 error 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: 404: Not Found],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
